pegasus-spark-bulkloader/src/main/java/com/xiaomi/infra/pegasus/spark/bulkloader/BulkLoader.java

```diff
@@ -0,0 +1,200 @@
+    }
+  }
+
+  private void createDataFile(Iterator<Tuple2<RocksDBRecord, String>> iterator)
+      throws PegasusException {
+    long start = System.currentTimeMillis();
+    long count = 0;
+
+    String curSSTFileName = curFileIndex + BULK_DATA_FILE_SUFFIX;
+    dataWriter.openWithRetry(partitionPath + curSSTFileName);
+    while (iterator.hasNext()) {
+      count++;
+      RocksDBRecord rocksDBRecord = iterator.next()._1;
+      if (curFileSize > SINGLE_FILE_SIZE_THRESHOLD) {
+        dataWriter.closeWithRetry();
+        LOG.debug(curFileIndex + BULK_DATA_FILE_SUFFIX + " writes complete!");
+
+        curFileIndex++;
+        curFileSize = 0L;
+        curSSTFileName = curFileIndex + BULK_DATA_FILE_SUFFIX;
+
+        dataWriter.openWithRetry(partitionPath + curSSTFileName);
+      }
+      curFileSize += dataWriter.writeWithRetry(rocksDBRecord.key(), rocksDBRecord.value());
+    }
+    if (curFileSize != 0) {
+      dataWriter.closeWithRetry();
+    } else {
+      // dataWriter will be throw exception when closed if dataWriter don't start any kv,
+      dataWriter.writeDefaultKV();
+      dataWriter.closeWithRetry();
+    }
+
+    LOG.info(
+        "create sst file time used is "
+            + (System.currentTimeMillis() - start)
+            + "ms, kv counts = "
+            + count
+            + " file counts = "
+            + curFileIndex);
+  }
+
+  private void createBulkLoadMetaDataFile() throws IOException, PegasusException {
+    long start = System.currentTimeMillis();
+    List<Future> taskList = new ArrayList<>();
+    AtomicBoolean isSuccess = new AtomicBoolean(true);
+
+    FileStatus[] fileStatuses = fdsService.getFileStatus(partitionPath);
+    for (FileStatus fileStatus : fileStatuses) {
+      taskList.add(
+          metaInfoCreateTask.submit(
+              () -> {
+                try {
+                  Tools.<Boolean>getDefaultRetryer().call(() -> generateFileMetaInfo(fileStatus));
+                } catch (ExecutionException | RetryException e) {
+                  isSuccess.set(false);
+                  LOG.error("write meta info[" + fileStatus.getPath().toString() + "] failed!");
+                }
+              }));
+    }
+
+    for (Future task : taskList) {
+      try {
+        task.get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new PegasusException("create metaDataInfo failed!", e);
+      }
+    }
+
+    if (!isSuccess.get()) {
+      throw new PegasusException("create metaDataInfo failed!");
+    }
+
+    dataMetaInfo.file_total_size = totalSize.get();
+    BufferedWriter bulkLoadMetaDataWriter = fdsService.getWriter(bulkLoadMetaDataPath);
+    bulkLoadMetaDataWriter.write(JSON.toJSONString(dataMetaInfo));
+    bulkLoadMetaDataWriter.close();
+    LOG.info("create meta info successfully, time used is " + (System.currentTimeMillis() - start));
+  }
+
+  private boolean generateFileMetaInfo(FileStatus fileStatus)
```

Comment:
This is a XiaoMi/FDS-specific method. Place it in FDSService, and make it a public interface.

We need a general interface of the remote filesystem abstraction, instead of binding FDS to BulkLoader. We could highly probably use other systems like HDFS in the future.

Reply:
`generateFileMetaInfo` is not `FDS-specific method`. `FDSService` has offered `common` method which can be access `hdfs` and `fds`. 

so `FDSService`  may need rename